Improved validation design

MarkPflug opened the pull request:
Yes, this is quite a bit more code, but the complexity is moved into the general-purpose ValidatingSchema.cs code, while your validation code becomes much simpler. 

I'm going to consider if/how I might move this ValidatingSchema into my libraries. But for now, I hope you find this design to be an improvement.

sherman89 reviewed (commented):
From the outside it looks very nice, but the code in `ValidatingSchema.cs` makes me worry a little bit about maintenance due to being slightly more complex...

If this stuff was hidden away in the library I would probably be happy, but I have to carefully consider now whether I want to put the maintenance burden of this on my team... I will still think about it and try to see if I can't refactor it but keep it easy to maintain and understand at the same time.

That being said I absolutely appreciate this PR, it does look very nice from the outside in Program.cs and I love the fact that the columns, their aliases, and validators are all defined in a single place 😄 

Hard to say whether this is the right design for the library, I'm not confident enough (yet) to say... probably some more eyes on this would be a good idea. I will try to think about it as soon as my wife lets me :-)

sherman89 commented:
Another requirement we have is to validate on row level, so for instance "If first name is not null, but last name is null, then row is not valid". Not sure which cell in this case would be considered invalid, I guess the latter one?

So basically I think it would be useful to be able to validate cells, columns, and rows, because some validations depend on other cells.

Maybe it's best to have a not-so-opinionated library but is it a crazy idea to emit events on every cell, row, and column read? That would flexibly allow the developer to validate for null on cell level, for duplicates on column level, and dependent cells on row level in this case.

Also would it be possible, just in-case, to anyway add invalid rows to results if one so wishes?

You already went above and beyond in helping me so I don't demand any help but maybe these are interesting thing to think about when designing validation for the library. Cheers!

sherman89 commented on SylvanExcelTest/Program.cs:
```diff
@@ -149,155 +95,81 @@ private static int Main()
+        foreach (var dbColumn in schema)
+        {
+            // If current columns do not contain expected column name, log error
+            if (excelSchema.All(c => c.BaseColumnName != dbColumn.BaseColumnName))
```
I just came back to this to think more about it and it seems like I'm taking the wrong approach here.. I cannot reliably expect the right column if I'm supporting multiple langauges unless it is decided beforehand either by the user or us.

So nevermind my previous comment 😄 

I will proceed to refactor with this implementation and see how it will turn out!

MarkPflug commented on SylvanExcelTest/ValidatingSchema.cs:
```diff
@@ -0,0 +1,268 @@
+                throw new ArgumentException();
+            }
+        }
+        foreach (var name in table.Names)
+        {
+            this.tableLookup.Add(name, table);
+        }
```
The reason for the duplicate loops is because the first pass is validating and the second pass is mutating. I wanted to avoid modifying the schema state if an exception was thrown. Ultimately, it probably doesn't matter, since these things will be constructed statically and recompile will be required to fix any issue. Additionally, it doesn't handle the case where the individual table has duplicate aliases.

MarkPflug commented:
I've been away on vacation, so I'm only seeing your feedback now.

I've added two changes:
1) fixed the ValidateSchema method to use the column Aliases.
2) added the BaseColumnName to the schema returned from the reader so it can be reported in error messages etc.

sherman89 commented:
@MarkPflug No worries, I already have a working version for the client using the "old style" of validation and it seems to work quite well. Maybe a few too many dictionaries but otherwise works well.

Regarding this PR, it looks cleaner, but I've come to some conclusions:

- Alises are fantastic, BUT, for multiple language support it seems like multiple tables is a better solution. Regardless of my requirements, it's probaby a nice addition.
  -  I don't remember all my reasonings (I'm a bit sick) but I guess mainly it was that I couldn't link an alias to a language to report possible errors. Kind of an edge case I know, but maybe indication that not appropriate usage in this case?
- The ability to easily pass external data to validators would be useful, since I need to transform a cell value into something else after validation, and the transformer class/func takes language as a parameter.
- UserValidator needs to know ordinals of FirstName and LastName, because I need to validate that they are both null or not together. There cannot be a first name without a last name and vice versa.

I couldn't push any changes here but basically those bullet points should hopefully be enough.

I will try to prepare a full solution with all my requirements and validations using the old flexible solution if you want to use it as a source of inspiration/ideas :)

Table syntax looks nice but for my needs it's a bit too inflexible compared to old ugly way I guess... Maybe Sylvan.Data.Excel should be as flexible and not-opinionated as possible. The current implementation is actually quite flexible which is great!

sherman89 commented:
@MarkPflug Whenever you feel like it, you can take a look at the master branch now (sorry forgot to make a new branch) if you'd like to see my approach to dealing with our clients requirements.

I'm not really happy with the code, too many dictionaries and worksheet validation alone made the code significantly uglier than it already was... I will think how to refactor but at the moment I cannot come up with anything better.

I think it was maybe a mistake for me to couple different language worksheets together as if they're one.. from a human perspective yes maybe, but they're actually different schemas.
